File: CodeHack2/src/com/codehack/tree/Fortress.java
package com.codehack.tree;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Scanner;

public class Fortress {
    private static Scanner sc;
    static int wallNum;
    static int[] x;
    static int[] y;
    static int[] r;
	private static int longest;
    
	public static void main(String[] args) {
        sc = new Scanner(System.in);
        int cases = sc.nextInt();
        
        while(cases-- > 0) {
        	longest = 0;
        	wallNum = sc.nextInt();
        	x = new int[wallNum];
        	y = new int[wallNum];
        	r = new int[wallNum];
        	
        	for (int i = 0; i < wallNum; i++) {
				x[i] = sc.nextInt();
				y[i] = sc.nextInt();
				r[i] = sc.nextInt();
			}
        	
        	TreeNode tree = getTree(0);
        	System.out.println( solve(tree) );
        }
	}

	private static int solve(TreeNode root) {
		longest = 0;
		int h = height( root );
		
		return Math.max(longest, h);
	}

	private static int height(TreeNode root) {
		ArrayList<Integer> heights = new ArrayList<Integer>();
		
		for (int i = 0; i < root.children.size(); i++) {
			heights.add( height(root.children.get(i)) );
		}
		
		if( heights.isEmpty() )	return 0;

		//	sort heights...
        Collections.sort( heights, new Comparator<Integer>(){
			@Override
			public int compare(Integer o1, Integer o2) {
				return o1 < o2 ? -1 : o1 > o2 ? 1 : 0;
			}
        });

		if( heights.size() >= 2 )
			longest = Math.max(longest, 
					2 + heights.get(heights.size()-2) + heights.get(heights.size()-1) );
		
		return heights.get( heights.size()-1 ) + 1;	//	get last
	}

	private static TreeNode getTree(int root) {
		TreeNode ret = new TreeNode();
		for (int i = 0; i < wallNum; i++) {
			if( isChild(root, i) )
				ret.children.add( getTree(i) );
		}
		
		return ret;
	}

	private static boolean isChild(int parent, int child) {
		if( !encloses(parent, child) )	return false;
		
		for (int i = 0; i < wallNum; i++) {
			if( i != parent && i != child &&
					encloses(parent, i) && encloses(i, child) )
				return false;
		}
		
		return true;
	}

	private static boolean encloses(int a, int b) {
		return r[a] > r[b] &&
				sqrtDist(a, b) < sqr(r[a] - r[b]);
	}

	private static int sqrtDist(int a, int b) {
		return sqr(y[a]-y[b]) + sqr(x[a]-x[b]);
	}

	private static int sqr(int x) {
		return x * x;
	}

}

class TreeNode	{
	ArrayList<TreeNode> children;
	
	public TreeNode()	{
		this.children = new ArrayList<TreeNode>();
	}
}

File: 21-implement_n_traverse_tree/fortress.py
# Problem ID: FORTRESS
# https://algospot.com/judge/problem/read/FORTRESS
# timeout

from collections import defaultdict

class Wall(object):
    def __init__(self):
        self.child = []


def readline():
    return raw_input()


def sqrdist(a, b, x_list, y_list):
    return ((y_list[a] - y_list[b])**2) + ((x_list[a] - x_list[b])**2)


def encloses(a, b, x_list, y_list, r_list):
    return r_list[a] > r_list[b] and sqrdist(a, b, x_list, y_list) < (r_list[a] - r_list[b])**2


def is_child(parent, child, x_list, y_list, r_list, n, child_dict):
    if child_dict[(parent,child)] != None:
        return child_dict[(parent,child)]
    result = None
    if not encloses(parent, child, x_list, y_list, r_list):
        result = child_dict[(parent,child)] = False
        return result
    for i in range(n):
        if i != parent and i != child and encloses(parent, i, x_list, y_list, r_list) and encloses(i, child, x_list, y_list, r_list):
            result = child_dict[(parent,child)] = False
            return result
    result = child_dict[(parent,child)] = True
    return result


def get_tree(root, x_list, y_list, r_list, n, child_dict):
    tree = Wall()
    for node in range(n):
        if is_child(root, node, x_list, y_list, r_list, n, child_dict):
            tree.child.append(get_tree(node, x_list, y_list, r_list, n, child_dict))
    return tree


def find_height(root_node, longest):
    heights = []
    for i in range(len(root_node.child)):
        heights.append(find_height(root_node.child[i], longest))
    if len(heights) == 0:
        return 0
    heights.sort()
    if len(heights) > 1:
        longest[0] = max(longest[0], 2 + heights[-2] + heights[-1])
    return heights[-1] + 1


def find_longest_path(root_node):
    longest = [0]
    height = find_height(root_node, longest)
    return max(longest[0], height)


def fortress():
    total_tests = int(readline())
    for testcase in range(total_tests):
        n = int(readline())
        x_list, y_list, r_list = [], [], []
        child_dict = defaultdict(lambda: None)
        for i in range(n):
            x, y, r = map(int, readline().split())
            x_list.append(x)
            y_list.append(y)
            r_list.append(r)
        root_node = get_tree(0, x_list, y_list, r_list, n, child_dict)
        print find_longest_path(root_node)


if __name__ == '__main__':
    fortress()  # run
